File: fastapi-app/app/services/todo_service.py
from sqlalchemy.orm import Session
from app.models.todo import Todo
from app.models.todo_log import TodoLog
from app.repositories import TodoRepository, TodoLogRepository
from app.exceptions import TodoTransactionError, TodoNotFoundError

class TodoService:
    """
    Business Logic Service Layer for managing Todo items and writing audit logs.
    
    This corresponds to com.example.todo.service.TodoService in the Spring Boot project.
    In Spring Boot, we annotate methods or the class with `@Transactional` to manage transactions.
    In FastAPI/SQLAlchemy, we use the `db.begin()` context manager to define transaction boundaries.
    """

    def __init__(self):
        self.todo_repo = TodoRepository()
        self.log_repo = TodoLogRepository()

    def get_all_todos(self, db: Session) -> list[Todo]:
        """Fetch all todos using the repository."""
        return self.todo_repo.find_all(db)

    def get_recent_logs(self, db: Session, limit: int = 50) -> list[TodoLog]:
        """Fetch recent transaction logs."""
        return self.log_repo.find_recent(db, limit)

    def create_todo(self, db: Session, title: str, description: str | None) -> Todo:
        """
        Creates a Todo and logs the action in a single database transaction.
        
        If an exception occurs within the `with db.begin()` block, SQLAlchemy
        automatically issues a ROLLBACK to the database. Otherwise, it issues a COMMIT.
        """
        # Defensive check
        if not title or not title.strip():
            raise ValueError("Todo title cannot be empty")

        todo = Todo(title=title, description=description)

        # Transaction boundary starts here
        with db.begin():
            # 1. Save the new Todo
            self.todo_repo.save(db, todo)
            db.flush()  # Generates the auto-increment ID in the database session
            
            # 2. Write the audit log
            log = TodoLog(
                todo_id=todo.id,
                action="CREATED",
                details=f"Created todo '{todo.title}'"
            )
            self.log_repo.save(db, log)

            # Simulated failure for testing transaction rollback
            if title == "force-rollback":
                raise TodoTransactionError("Simulated failure: forcing database transaction rollback")
        
        # Transaction is committed and connection closed here (outside the block)
        return todo

    def toggle_todo_status(self, db: Session, todo_id: int) -> Todo:
        """Toggles the completed status of a Todo and logs the action in a transaction."""
        with db.begin():
            todo = self.todo_repo.find_by_id(db, todo_id)
            if not todo:
                raise TodoNotFoundError(f"Todo with ID {todo_id} not found")

            # Toggle status
            todo.completed = not todo.completed
            self.todo_repo.save(db, todo)

            action_desc = "COMPLETED" if todo.completed else "INCOMPLETE"
            log = TodoLog(
                todo_id=todo.id,
                action="STATUS_UPDATED",
                details=f"Marked todo '{todo.title}' as {action_desc}"
            )
            self.log_repo.save(db, log)

        return todo

    def delete_todo(self, db: Session, todo_id: int) -> None:
        """Deletes a Todo and logs the action in a transaction."""
        with db.begin():
            todo = self.todo_repo.find_by_id(db, todo_id)
            if not todo:
                raise TodoNotFoundError(f"Todo with ID {todo_id} not found")

            self.todo_repo.delete(db, todo)

            log = TodoLog(
                todo_id=todo_id,
                action="DELETED",
                details=f"Deleted todo '{todo.title}'"
            )
            self.log_repo.save(db, log)

File: springboot-app/src/main/java/com/example/todo/service/TodoService.java
package com.example.todo.service;

import com.example.todo.model.Todo;
import com.example.todo.model.TodoLog;
import com.example.todo.repository.TodoLogRepository;
import com.example.todo.repository.TodoRepository;
import com.example.todo.exception.TodoNotFoundException;
import com.example.todo.exception.TodoTransactionException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Service class for handling all business logic related to Todos and Todo Logs.
 *
 * <p><strong>Python/FastAPI Comparison:</strong>
 * In FastAPI, transaction control is often manual using SQLAlchemy's context managers
 * ({@code with session.begin():} or {@code try/except} blocks with {@code session.rollback()}).
 * In Spring Boot, transaction management is declarative. By annotating the service class or methods
 * with {@link Transactional}, Spring automatically opens a transaction, executes the method code,
 * and commits it. If a runtime exception is thrown, Spring automatically rolls back all database operations
 * executed within that transaction boundary.
 * </p>
 */
@Service
@Transactional
public class TodoService {

    private final TodoRepository todoRepository;
    private final TodoLogRepository todoLogRepository;

    public TodoService(TodoRepository todoRepository, TodoLogRepository todoLogRepository) {
        this.todoRepository = todoRepository;
        this.todoLogRepository = todoLogRepository;
    }

    /**
     * Retrieve all Todo items.
     *
     * @return a list of all Todos
     */
    @Transactional(readOnly = true)
    public List<Todo> getAllTodos() {
        return todoRepository.findAll();
    }

    /**
     * Retrieve a specific Todo item by ID.
     *
     * @param id the ID of the Todo
     * @return the Todo if found, or empty
     */
    @Transactional(readOnly = true)
    public Optional<Todo> getTodoById(Long id) {
        return todoRepository.findById(id);
    }

    /**
     * Retrieve all log records.
     *
     * @return list of all TodoLog entries
     */
    @Transactional(readOnly = true)
    public List<TodoLog> getAllLogs() {
        return todoLogRepository.findAllByOrderByTimestampDesc();
    }

    /**
     * Create a new Todo item and write an audit log entry.
     * Triggers a rollback if the title is "force-rollback".
     *
     * @param todo the Todo to create
     * @return the created Todo
     */
    public Todo createTodo(Todo todo) {
        // Save the Todo first to generate an ID
        Todo savedTodo = todoRepository.save(todo);

        // Write a log entry
        TodoLog log = new TodoLog(
            savedTodo.getId(),
            "CREATE",
            "Created Todo: '" + savedTodo.getTitle() + "' (Completed: " + savedTodo.isCompleted() + ")"
        );
        todoLogRepository.save(log);

        // Check for force-rollback trigger
        if ("force-rollback".equalsIgnoreCase(savedTodo.getTitle())) {
            throw new TodoTransactionException("Forced transaction rollback triggered by title: 'force-rollback'");
        }

        return savedTodo;
    }

    /**
     * Update an existing Todo item and write an audit log entry.
     * Triggers a rollback if the new title is "force-rollback".
     *
     * @param id the ID of the Todo to update
     * @param updatedDetails the new details for the Todo
     * @return the updated Todo
     */
    public Todo updateTodo(Long id, Todo updatedDetails) {
        Todo existingTodo = todoRepository.findById(id)
            .orElseThrow(() -> new TodoNotFoundException("Todo with ID " + id + " not found"));

        String oldTitle = existingTodo.getTitle();
        boolean oldCompleted = existingTodo.isCompleted();

        // Update fields
        existingTodo.setTitle(updatedDetails.getTitle());
        existingTodo.setDescription(updatedDetails.getDescription());
        existingTodo.setCompleted(updatedDetails.isCompleted());

        Todo savedTodo = todoRepository.save(existingTodo);

        // Write log entry
        TodoLog log = new TodoLog(
            savedTodo.getId(),
            "UPDATE",
            String.format("Updated Todo ID %d. Title: '%s' -> '%s', Completed: %s -> %s",
                id, oldTitle, savedTodo.getTitle(), oldCompleted, savedTodo.isCompleted())
        );
        todoLogRepository.save(log);

        // Check for force-rollback trigger
        if ("force-rollback".equalsIgnoreCase(savedTodo.getTitle())) {
            throw new TodoTransactionException("Forced transaction rollback triggered by title: 'force-rollback'");
        }

        return savedTodo;
    }

    /**
     * Delete a Todo item and write an audit log entry.
     *
     * @param id the ID of the Todo to delete
     */
    public void deleteTodo(Long id) {
        Todo existingTodo = todoRepository.findById(id)
            .orElseThrow(() -> new TodoNotFoundException("Todo with ID " + id + " not found"));

        // Write log entry
        TodoLog log = new TodoLog(
            id,
            "DELETE",
            "Deleted Todo: '" + existingTodo.getTitle() + "'"
        );
        todoLogRepository.save(log);

        // Delete the Todo
        todoRepository.delete(existingTodo);
    }

    /**
     * Clears all logs in the database.
     */
    public void clearLogs() {
        todoLogRepository.deleteAll();
    }
}
